src/main/java/org/opensearch/knn/index/codec/nativeindex/remote/RemoteIndexBuildStrategy.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.nativeindex.remote;
+
+import lombok.extern.log4j.Log4j2;
+import org.apache.commons.lang.NotImplementedException;
+import org.apache.lucene.index.SegmentWriteState;
+import org.opensearch.common.StopWatch;
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.index.IndexSettings;
+import org.opensearch.knn.common.featureflags.KNNFeatureFlags;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.index.codec.nativeindex.NativeIndexBuildStrategy;
+import org.opensearch.knn.index.codec.nativeindex.model.BuildIndexParams;
+import org.opensearch.knn.index.vectorvalues.KNNVectorValues;
+import org.opensearch.repositories.RepositoriesService;
+import org.opensearch.repositories.Repository;
+import org.opensearch.repositories.RepositoryMissingException;
+import org.opensearch.repositories.blobstore.BlobStoreRepository;
+
+import java.io.IOException;
+import java.util.function.Supplier;
+
+import static org.opensearch.knn.index.KNNSettings.KNN_INDEX_REMOTE_VECTOR_BUILD_SETTING;
+import static org.opensearch.knn.index.KNNSettings.KNN_REMOTE_VECTOR_REPO_SETTING;
+
+/**
+ * This class orchestrates building vector indices. It handles uploading data to a repository, submitting a remote
+ * build request, awaiting upon the build request to complete, and finally downloading the data from a repository.
+ */
+@Log4j2
+@ExperimentalApi
+public class RemoteIndexBuildStrategy implements NativeIndexBuildStrategy {
+
+    private final Supplier<RepositoriesService> repositoriesServiceSupplier;
+    private final NativeIndexBuildStrategy fallbackStrategy;
+    private static final String VECTOR_BLOB_FILE_EXTENSION = ".knnvec";
+    private static final String DOC_ID_FILE_EXTENSION = ".knndid";
+
+    /**
+     * Public constructor
+     *
+     * @param repositoriesServiceSupplier   A supplier for {@link RepositoriesService} used for interacting with repository
+     */
+    public RemoteIndexBuildStrategy(Supplier<RepositoriesService> repositoriesServiceSupplier, NativeIndexBuildStrategy fallbackStrategy) {
+        this.repositoriesServiceSupplier = repositoriesServiceSupplier;
+        this.fallbackStrategy = fallbackStrategy;
+    }
+
+    /**
+     * @return whether to use the remote build feature
+     */
+    public static boolean shouldBuildIndexRemotely(IndexSettings indexSettings) {
+        String vectorRepo = KNNSettings.state().getSettingValue(KNN_REMOTE_VECTOR_REPO_SETTING.getKey());
+        return KNNFeatureFlags.isKNNRemoteVectorBuildEnabled()
+            && indexSettings.getValue(KNN_INDEX_REMOTE_VECTOR_BUILD_SETTING)
+            && vectorRepo != null
+            && !vectorRepo.isEmpty();
+    }
```

Comment:
This logic needs to be involved bit more where the doc count needs to be added, say if number of docs in segment < 10K we should go with local build as compare to remote build. This number 10K can be tuned, or we can make this logic more sophisticated.

So please add that check in the next PR.